File: leet-code/medium/cpp/0002 add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
  ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    int rd = 0;
    ListNode* result = l1;
    ListNode* tmp;
    
    while (l1 || l2) {
      tmp = l1;
      if (!tmp) {
        tmp = l2;
      } else if (l2) {
        tmp->val += l2->val;
      }
      
      tmp->val += rd;
      
      rd = tmp->val / 10;
      tmp->val %= 10;
      
      if (!l1->next && (l2 && l2->next)) {
        l1->next = l2->next;
        l2->next = nullptr;
      }
      
      l1 = this->getNext(l1);
      l2 = this->getNext(l2);
    }
    
    if (rd) {
      tmp->next = new ListNode(rd);
    }
    
    return result;
  };
  
  ListNode* getNext(ListNode* l) {
    return l && l->next ? l->next : nullptr;
  }
};

File: leet-code/python/0002_add_two_numbers.py
# Definition for singly-linked list.
class ListNode:
  def __init__(self, val=0, next=None):
    self.val = val
    self.next = next
    
class Solution:
  def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    carry = False
    rr = ListNode()
    cr = rr
    while l1 is not None:
      if carry:
        cr.val += 1
        carry = False
        
      cr.val += l1.val
      if l2 is not None:
        cr.val += l2.val
      
      if cr.val >= 10:
        carry = True
        cr.val = cr.val % 10
      
      l1 = l1.next
      if l2 is not None:
        l2 = l2.next
      
      if l1 is None and l2 is not None:
        l1 = l2
        l2 = None
      
      if l1 is None:
        break
      
      cr.next = ListNode()
      cr = cr.next
      
    if carry:
      cr.next = ListNode(1)
      
    return rr
        